cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1286,700 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+  if (splits.size() > 0) {
+    CUDF_EXPECTS(splits.back() <= input.column(0).size(),
+                 "splits can't exceed size of input columns");
+  }
+  {
+    size_type begin = 0;
+    for (std::size_t i = 0; i < splits.size(); i++) {
+      size_type end = splits[i];
+      CUDF_EXPECTS(begin >= 0, "Starting index cannot be negative.");
+      CUDF_EXPECTS(end >= begin, "End index cannot be smaller than the starting index.");
+      CUDF_EXPECTS(end <= input.column(0).size(), "Slice range out of bounds.");
+      begin = end;
+    }
+  }
+  return input.column(0).size() == 0;
+}
+
+/**
+ * @brief A helper struct containing the state of contiguous_split, whether the caller
+ * is using the single-pass contiguous_split or chunked_pack.
+ *
+ * It exposes an iterator-like pattern where contiguous_split_state::has_next()
+ * return true when there is work to be done, and false otherwise.
+ *
+ * contiguous_split_state::contiguous_split() performs a single-pass contiguous_split
+ * and is only valid iff contiguous_split_state is instantiated with 0 for the user_buffer_size.
+ *
+ * contiguous_split_state::contiguous_split_chunk(device_span) is only valid when
+ * user_buffer_size > 0. It should be called as long as has_next() returns true. The
+ * device_span passed to contiguous_split_chunk must be allocated in stream `stream` by
+ * the user.
+ *
+ * None of the methods are thread safe.
+ */
+struct contiguous_split_state {
+  static const std::size_t desired_batch_size = 1 * 1024 * 1024;
+
+  contiguous_split_state(cudf::table_view const& input,
+                         std::size_t user_buffer_size,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr)
+    : contiguous_split_state(input, {}, user_buffer_size, stream, mr)
+  {
+  }
+
+  contiguous_split_state(cudf::table_view const& input,
+                         std::vector<size_type> const& splits,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr)
+    : contiguous_split_state(input, splits, 0, stream, mr)
+  {
+  }
+
+  contiguous_split_state(cudf::table_view const& input,
+                         std::vector<size_type> const& splits,
+                         std::size_t user_buffer_size,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr)
+    : input(input), user_buffer_size(user_buffer_size), stream(stream), mr(mr)
+  {
+    is_empty       = check_inputs(input, splits);
+    num_partitions = splits.size() + 1;
+    num_src_bufs   = count_src_bufs(input.begin(), input.end());
+    num_bufs       = num_src_bufs * num_partitions;
+
+    // if the table we are about to contig split is empty, we have special
+    // handling where metadata is produced and a 0-byte contiguous buffer
+    // is the result.
+    if (is_empty) { return; }
+
+    partition_buf_size_and_dst_buf_info =
```

Comment:
This module is huge, but a big part of it is all (the setup anyway) expressed in these 4 operations.  Off the top of my head I don't have any bright ideas but I'd love to have a way of putting a big bright neon sign pointing at this being the key.